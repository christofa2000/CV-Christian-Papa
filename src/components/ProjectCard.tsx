"use client";

import { motion, useAnimation, useInView } from "framer-motion";
import { ExternalLink, Github } from "lucide-react";
import Image from "next/image";
import { useEffect, useRef } from "react";

interface Project {
  id: string;
  title: string;
  description: string;
  image: string;
  demoUrl?: string;
  repoUrl?: string;
  technologies: string[];
  highlights?: string[];
  // side?: "left" | "right"; // â† ya no lo usamos para el layout
}

interface ProjectCardProps {
  project: Project;
  index: number;
}

export default function ProjectCard({ project, index }: ProjectCardProps) {
  // Alternado por Ã­ndice: pares => imagen izquierda; impares => imagen derecha
  const isLeft = index % 2 === 0;

  const rootRef = useRef<HTMLDivElement>(null);
  const inView = useInView(rootRef, { amount: 0.3 });
  const controls = useAnimation();

  useEffect(() => {
    controls.start(inView ? "show" : "hidden");
  }, [inView, controls]);

  // Imagen entra desde su borde natural
  const imageSlide = {
    hidden: { x: isLeft ? -220 : 220, opacity: 0 },
    show: {
      x: 0,
      opacity: 1,
      transition: { type: "spring", stiffness: 70, damping: 14 },
    },
  };

  // Texto entra desde el lado opuesto
  const textSlide = {
    hidden: { x: isLeft ? 220 : -220, opacity: 0 },
    show: {
      x: 0,
      opacity: 1,
      transition: { type: "spring", stiffness: 70, damping: 14, delay: 0.03 },
    },
  };

  return (
    <motion.article
      ref={rootRef}
      initial="hidden"
      animate={controls}
      viewport={{ once: false, amount: 0.3 }}
      className="grid lg:grid-cols-2 gap-12 items-center"
    >
      {/* IMAGEN */}
      <motion.div
        variants={imageSlide}
        className={`group relative aspect-[16/10] rounded-2xl overflow-hidden border border-white/10 shadow-lg
          ${
            isLeft ? "lg:order-1" : "lg:order-2"
          } w-full lg:w-4/5 xl:w-3/4 mx-auto`}
      >
        <Image
          src={project.image}
          alt={`Screenshot del proyecto ${project.title}`}
          fill
          className="object-cover transition-transform duration-300 group-hover:scale-[1.03]"
          sizes="(max-width: 768px) 100vw, 50vw"
        />

        {/* Overlay con botones */}
        <div className="absolute inset-0 bg-black/60 opacity-0 group-hover:opacity-100 transition-opacity duration-300 flex items-center justify-center">
          <div className="flex gap-4">
            {project.demoUrl && (
              <a
                href={project.demoUrl}
                target="_blank"
                rel="noopener noreferrer"
                className="p-3 bg-white text-neutral-900 rounded-full hover:bg-blue-700 hover:text-white transition-colors focus-ring"
                aria-label={`Ver demo de ${project.title}`}
              >
                <ExternalLink size={20} />
              </a>
            )}
            {project.repoUrl && (
              <a
                href={project.repoUrl}
                target="_blank"
                rel="noopener noreferrer"
                className="p-3 bg-white text-neutral-900 rounded-full hover:bg-neutral-800 hover:text-white transition-colors focus-ring"
                aria-label={`Ver Código de ${project.title}`}
              >
                <Github size={20} />
              </a>
            )}
          </div>
        </div>
      </motion.div>

      {/* TEXTO */}
      <motion.div
        variants={textSlide}
        className={`space-y-6 ${
          isLeft ? "lg:order-2" : "lg:order-1"
        } w-full lg:w-4/5 xl:w-3/4 mx-auto`}
      >
        <div className="space-y-4">
          <h3 className="text-3xl font-semibold text-neutral-200">
            {project.title}
          </h3>
          <p className="text-lg text-neutral-400 leading-relaxed">
            {project.description}
          </p>
        </div>

        <div className="flex flex-wrap gap-2">
          {project.technologies.map((tech) => (
            <span
              key={tech}
              className="px-3 py-1 bg-neutral-800 text-neutral-300 rounded-full text-sm font-medium border border-white/10"
            >
              {tech}
            </span>
          ))}
        </div>

        <div className="flex flex-wrap gap-4">
          {project.demoUrl && (
            <a
              href={project.demoUrl}
              target="_blank"
              rel="noopener noreferrer"
              className="inline-flex items-center gap-2 px-6 py-3 bg-blue-700 text-white font-medium rounded-xl hover:bg-blue-600 transition-colors focus-ring"
            >
              <ExternalLink size={16} />
              Ver Demo
            </a>
          )}
          {project.repoUrl && (
            <a
              href={project.repoUrl}
              target="_blank"
              rel="noopener noreferrer"
              className="inline-flex items-center gap-2 px-6 py-3 border border-neutral-600 text-neutral-300 font-medium rounded-xl hover:bg-neutral-800 transition-colors focus-ring"
            >
              <Github size={16} />
              Código
            </a>
          )}
        </div>
      </motion.div>
    </motion.article>
  );
}
