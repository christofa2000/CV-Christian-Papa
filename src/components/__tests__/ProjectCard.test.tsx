import { render, screen } from "@testing-library/react";
import ProjectCard from "../ProjectCard";

const mockProject = {
  id: "test-project",
  title: "Proyecto de Prueba",
  description: "Una descripción de prueba para el proyecto",
  image: "/placeholder-project-1.jpg",
  demoUrl: "https://demo.example.com",
  repoUrl: "https://github.com/example/repo",
  technologies: ["React", "TypeScript", "Next.js"],
};

describe("ProjectCard component", () => {
  it("renderiza el título del proyecto", () => {
    render(<ProjectCard project={mockProject} index={0} />);
    expect(screen.getByText(/Proyecto de Prueba/i)).toBeInTheDocument();
  });

  it("renderiza la descripción del proyecto", () => {
    render(<ProjectCard project={mockProject} index={0} />);
    expect(
      screen.getByText(/Una descripción de prueba para el proyecto/i)
    ).toBeInTheDocument();
  });

  it("renderiza las tecnologías como tags", () => {
    render(<ProjectCard project={mockProject} index={0} />);
    expect(screen.getByText(/React/i)).toBeInTheDocument();
    expect(screen.getByText(/TypeScript/i)).toBeInTheDocument();
    expect(screen.getByText(/Next.js/i)).toBeInTheDocument();
  });

  it("renderiza los botones de demo y código", () => {
    render(<ProjectCard project={mockProject} index={0} />);
    expect(screen.getByText(/Ver Demo/i)).toBeInTheDocument();
    expect(screen.getByText(/Código/i)).toBeInTheDocument();
  });

  it("renderiza la imagen con el alt text correcto", () => {
    render(<ProjectCard project={mockProject} index={0} />);
    const image = screen.getByAltText(
      /Screenshot del proyecto Proyecto de Prueba/i
    );
    expect(image).toBeInTheDocument();
  });

  it("tiene los enlaces con los hrefs correctos", () => {
    render(<ProjectCard project={mockProject} index={0} />);
    const demoLink = screen.getByText(/Ver Demo/i).closest("a");
    const codeLink = screen.getByText(/Código/i).closest("a");

    expect(demoLink).toHaveAttribute("href", "https://demo.example.com");
    expect(codeLink).toHaveAttribute("href", "https://github.com/example/repo");
  });

  it("coincide con el snapshot", () => {
    const { container } = render(
      <ProjectCard project={mockProject} index={0} />
    );
    expect(container).toMatchSnapshot();
  });
});


