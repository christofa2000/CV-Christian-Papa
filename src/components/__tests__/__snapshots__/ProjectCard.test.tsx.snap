// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProjectCard component coincide con el snapshot 1`] = `
<div>
  <article
    class="grid lg:grid-cols-2 gap-12 items-center"
  >
    <div
      class="group relative aspect-[16/10] rounded-2xl overflow-hidden border border-white/10 shadow-lg
          lg:order-1 w-full lg:w-4/5 xl:w-3/4 mx-auto"
    >
      <img
        alt="Screenshot del proyecto Proyecto de Prueba"
        class="object-cover transition-transform duration-300 group-hover:scale-[1.03]"
        src="/placeholder-project-1.jpg"
      />
      <div
        class="absolute inset-0 bg-black/60 opacity-0 group-hover:opacity-100 transition-opacity duration-300 flex items-center justify-center"
      >
        <div
          class="flex gap-4"
        >
          <a
            aria-label="Ver demo de Proyecto de Prueba"
            class="p-3 bg-white text-neutral-900 rounded-full hover:bg-indigo-600 hover:text-white transition-colors focus-ring"
            href="https://demo.example.com"
            rel="noopener noreferrer"
            target="_blank"
          >
            <svg
              class="lucide lucide-external-link"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15 3h6v6"
              />
              <path
                d="M10 14 21 3"
              />
              <path
                d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
              />
            </svg>
          </a>
          <a
            aria-label="Ver Código de Proyecto de Prueba"
            class="p-3 bg-white text-neutral-900 rounded-full hover:bg-neutral-800 hover:text-white transition-colors focus-ring"
            href="https://github.com/example/repo"
            rel="noopener noreferrer"
            target="_blank"
          >
            <svg
              class="lucide lucide-github"
              fill="none"
              height="20"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15 22v-4a4.8 4.8 0 0 0-1-3.5c3 0 6-2 6-5.5.08-1.25-.27-2.48-1-3.5.28-1.15.28-2.35 0-3.5 0 0-1 0-3 1.5-2.64-.5-5.36-.5-8 0C6 2 5 2 5 2c-.3 1.15-.3 2.35 0 3.5A5.403 5.403 0 0 0 4 9c0 3.5 3 5.5 6 5.5-.39.49-.68 1.05-.85 1.65-.17.6-.22 1.23-.15 1.85v4"
              />
              <path
                d="M9 18c-4.51 2-5-2-7-2"
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
    <div
      class="space-y-6 lg:order-2 w-full lg:w-4/5 xl:w-3/4 mx-auto"
    >
      <div
        class="space-y-4"
      >
        <h3
          class="text-3xl font-semibold text-neutral-200"
        >
          Proyecto de Prueba
        </h3>
        <p
          class="text-lg text-neutral-400 leading-relaxed"
        >
          Una descripción de prueba para el proyecto
        </p>
      </div>
      <div
        class="flex flex-wrap gap-2"
      >
        <span
          class="px-3 py-1 bg-neutral-800 text-neutral-300 rounded-full text-sm font-medium border border-white/10"
        >
          React
        </span>
        <span
          class="px-3 py-1 bg-neutral-800 text-neutral-300 rounded-full text-sm font-medium border border-white/10"
        >
          TypeScript
        </span>
        <span
          class="px-3 py-1 bg-neutral-800 text-neutral-300 rounded-full text-sm font-medium border border-white/10"
        >
          Next.js
        </span>
      </div>
      <div
        class="flex flex-wrap gap-4"
      >
        <a
          class="inline-flex items-center gap-2 px-6 py-3 bg-indigo-600 text-white font-medium rounded-xl hover:bg-indigo-700 transition-colors focus-ring"
          href="https://demo.example.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            class="lucide lucide-external-link"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 3h6v6"
            />
            <path
              d="M10 14 21 3"
            />
            <path
              d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
            />
          </svg>
          Ver Demo
        </a>
        <a
          class="inline-flex items-center gap-2 px-6 py-3 border border-neutral-600 text-neutral-300 font-medium rounded-xl hover:bg-neutral-800 transition-colors focus-ring"
          href="https://github.com/example/repo"
          rel="noopener noreferrer"
          target="_blank"
        >
          <svg
            class="lucide lucide-github"
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 22v-4a4.8 4.8 0 0 0-1-3.5c3 0 6-2 6-5.5.08-1.25-.27-2.48-1-3.5.28-1.15.28-2.35 0-3.5 0 0-1 0-3 1.5-2.64-.5-5.36-.5-8 0C6 2 5 2 5 2c-.3 1.15-.3 2.35 0 3.5A5.403 5.403 0 0 0 4 9c0 3.5 3 5.5 6 5.5-.39.49-.68 1.05-.85 1.65-.17.6-.22 1.23-.15 1.85v4"
            />
            <path
              d="M9 18c-4.51 2-5-2-7-2"
            />
          </svg>
          Código
        </a>
      </div>
    </div>
  </article>
</div>
`;
